conn.go

```diff
@@ -2,50 +2,89 @@ package conn
 
 import (
 	"context"
+	"errors"
 	"io"
 	"net"
-	"time"
 
 	logging "github.com/ipfs/go-log"
-	mpool "github.com/jbenet/go-msgio/mpool"
+	ci "github.com/libp2p/go-libp2p-crypto"
 	ic "github.com/libp2p/go-libp2p-crypto"
 	iconn "github.com/libp2p/go-libp2p-interface-conn"
 	lgbl "github.com/libp2p/go-libp2p-loggables"
 	peer "github.com/libp2p/go-libp2p-peer"
+	secio "github.com/libp2p/go-libp2p-secio"
 	tpt "github.com/libp2p/go-libp2p-transport"
+	smux "github.com/libp2p/go-stream-muxer"
 	ma "github.com/multiformats/go-multiaddr"
 )
 
 var log = logging.Logger("conn")
 
-// ReleaseBuffer puts the given byte array back into the buffer pool,
-// first verifying that it is the correct size
-func ReleaseBuffer(b []byte) {
-	log.Debugf("Releasing buffer! (cap,size = %d, %d)", cap(b), len(b))
-	mpool.ByteSlicePool.Put(uint32(cap(b)), b)
-}
-
-// singleConn represents a single connection to another Peer (IPFS Node).
+// singleConn represents a single stream-multipexed connection to another Peer (IPFS Node).
 type singleConn struct {
-	local  peer.ID
-	remote peer.ID
-	maconn tpt.Conn
-	event  io.Closer
+	streamConn smux.Conn
+	tptConn    tpt.Conn
+
+	secSession secio.Session
+
+	event io.Closer
 }
 
-// newConn constructs a new connection
-func newSingleConn(ctx context.Context, local, remote peer.ID, maconn tpt.Conn) (iconn.Conn, error) {
-	ml := lgbl.Dial("conn", local, remote, maconn.LocalMultiaddr(), maconn.RemoteMultiaddr())
+var _ iconn.Conn = &singleConn{}
 
-	conn := &singleConn{
-		local:  local,
-		remote: remote,
-		maconn: maconn,
-		event:  log.EventBegin(ctx, "connLifetime", ml),
+// newSingleConn constructs a new connection
+func newSingleConn(ctx context.Context, local, remote peer.ID, privKey ci.PrivKey, tptConn tpt.Conn, pstpt smux.Transport, isServer bool) (iconn.Conn, error) {
+	ml := lgbl.Dial("conn", local, remote, tptConn.LocalMultiaddr(), tptConn.RemoteMultiaddr())
+
+	var streamConn smux.Conn
+	var secSession secio.Session
+
+	c := tptConn
+	// 1. secure the connection
+	if privKey != nil && iconn.EncryptConnections {
+		var err error
+		secSession, err = setupSecureSession(ctx, local, privKey, tptConn)
+		if err != nil {
+			return nil, err
+		}
+		c = &secureConn{
+			insecure: tptConn,
+			secure:   secSession,
+		}
```

Comment:
If `remote` is not `""`, we should check if `remote == secSession.RemoteID()` here. That way, we do this *before* doing anything else. Currently, we check in swarm but, IMO, that's too late. Note: this is an improvement, not a bug.

Reply:
Can we do that in a separate PR?

Reply:
Fair enough.